engine/render/src/render/render.cpp

```diff
@@ -828,17 +828,29 @@ namespace dmRender
 
             ApplyRenderState(render_context, render_context->m_GraphicsContext, dmGraphics::GetPipelineState(context), ro);
 
+            uint8_t next_texture_unit = 0;
             for (uint32_t i = 0; i < RenderObject::MAX_TEXTURE_COUNT; ++i)
             {
                 dmGraphics::HTexture texture = ro->m_Textures[i];
                 if (render_context->m_Textures[i])
                     texture = render_context->m_Textures[i];
+
                 if (texture)
                 {
-                    dmGraphics::EnableTexture(context, i, texture);
-                    ApplyMaterialSampler(render_context, material, i, texture);
+                    HSampler sampler = GetMaterialSampler(material, i);
+                    if (GetCanBindTexture(texture, sampler, i))
+                    {
+                        // NOTE: in the case of texture arrays on platforms that doesn't support them,
+                        //       we need to separate between texture units and samplers. Otherwise,
+                        //       we would need separate texture samplers for every sub-texture.
+                        for (int sub_handle = 0; sub_handle < dmGraphics::GetNumTextureHandles(texture); ++sub_handle)
+                        {
+                            dmGraphics::EnableTexture(context, next_texture_unit, sub_handle, texture);
+                            ApplyMaterialSampler(render_context, material, sampler, next_texture_unit, texture);
+                            next_texture_unit++;
```

Comment:
There might be a bit of a disconnect here between our "MAX_TEXTURE_COUNT" value. The fallback version will "eat" x actual sampler slots with this loop meaning we could try to bind more than the max value. E.g a render object has 4 texture it wants to bind and if two of those are array textures with 4 slots each we will surpass the max count. I think this is the same case with an actual array texture as well, so we might want to deal with this in some other way, maybe query the GPU for how many textures we are allowed to bind? The MAX_TEXTURE_COUNT value is based on the lowest number of supported samplers in the GL2 spec afaik..